pallets/gear-builtin/src/lib.rs

```diff
@@ -0,0 +1,368 @@
+            registry.entry(actor_id).or_insert(builtin_id);
+        }
+        assert!(
+            registry.len() == builtin_ids_len,
+            "Duplicate builtin ids detected!"
+        );
+
+        Self {
+            registry,
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<T: Config> BuiltinRouter for BuiltinRegistry<T> {
+    type QueuedDispatch = StoredDispatch;
+    type Output = JournalNote;
+
+    fn lookup(&self, id: &ProgramId) -> Option<BuiltinId> {
+        self.registry.get(id).copied()
+    }
+
+    fn dispatch(&self, dispatch: StoredDispatch, gas_limit: u64) -> Option<Vec<JournalNote>> {
+        let actor_id = dispatch.destination();
+
+        // Try to convert the incoming dispatch into a format that the builtin actor can handle.
+        let builtin_message =
+            <T::Message as TryFrom<StoredDispatch>>::try_from(dispatch.clone()).ok()?;
+
+        // Builtin actors can only execute dispatches of `Handle` kind and only `Handle`
+        // dispatches can end up here (TODO: elaborate).
+        if dispatch.kind() != DispatchKind::Handle {
+            unreachable!("Only handle dispatches can end up here");
+        }
+
+        let mut gas_counter = GasCounter::new(gas_limit);
+
+        let (res, gas_spent) = <T as Config>::BuiltinActor::handle(&builtin_message, gas_limit);
+
+        // We rely on a builtin actor having performed the check for gas limit consistency
+        // and having reported an error if the `gas_limit` was to have been exceeded.
+        // However, to avoid gas tree corruption error, we must not report as spent more gas than
+        // the amount reserved in gas tree (that is, `gas_limit`). Hence (just in case):
+        let gas_spent = gas_spent.min(gas_limit);
+
+        // This should always return `ChargeResult::Enough` now thanks to the above check.
+        let _ = gas_counter.charge(gas_spent);
+
+        let dispatch = dispatch.into_incoming(gas_limit);
+
+        match res {
+            Ok(response_bytes) => {
+                log::debug!(target: LOG_TARGET, "Builtin call dispatched successfully");
+
+                let mut dispatch_result =
+                    DispatchResult::success(dispatch.clone(), actor_id, gas_counter.to_amount());
+
+                // Create an artificial `MessageContext` object that will help us to generate
+                // a reply from the builtin actor.
+                let mut message_context =
+                    MessageContext::new(dispatch, actor_id, Default::default());
+
+                let payload = response_bytes
+                    .as_ref()
+                    .to_vec()
+                    .try_into()
+                    .unwrap_or_else(|_| unreachable!("Response message is too large"));
+                let packet = ReplyPacket::new(payload, 0);
+
+                // Mark reply as sent
+                if let Ok(_reply_id) = message_context.reply_commit(packet.clone(), None) {
```

Comment:
Error is virtually impossible here: it can only occur if there are some previous reply chunks already being stored for this dispatch, so when you try to recollect them all together they might exceed the size of the `LimitedVec`. But in our case we only have one packet.
Even if there was an error (again, impossible case) I guess we would just have ended up with auto-reply instead of a reply message with the payload generated by the builtin actor.